File: cert-processor/src/main/java/org/sunbird/incredible/processor/qrcode/AccessCodeGenerator.java
package org.sunbird.incredible.processor.qrcode;


import org.apache.commons.lang3.StringUtils;

import java.math.BigDecimal;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AccessCodeGenerator {

    private String stripChars = "0";
    private Double length = 6.0;
    private BigDecimal largePrimeNumber = new BigDecimal(1679979167);
    private String regex = "[A-Z][0-9][A-Z][0-9][A-Z][0-9]";
    private Pattern pattern = Pattern.compile(regex);

    private static final String[] ALPHABET = new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C",
            "D", "E", "F", "G", "H", "J", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y",
            "Z"};

    public AccessCodeGenerator(Double length) {
        this.length = length;
    }

    public String generate() {
        double count = 1;
        int totalChars = ALPHABET.length;
        BigDecimal exponent = BigDecimal.valueOf(totalChars);
        exponent = exponent.pow(length.intValue());
        double codesCount = 0;
        double lastIndex = 0;
        String code = null;
        while (codesCount < count) {
            lastIndex = getMaxIndex();
            BigDecimal number = new BigDecimal(lastIndex);
            BigDecimal num = number.multiply(largePrimeNumber).remainder(exponent);
            code = baseN(num, totalChars);
            if (code.length() == length && isValidCode(code)) {
                codesCount += 1;
            }
        }
        return code;

    }

    private String baseN(BigDecimal num, int base) {
        if (num.doubleValue() == 0) {
            return "0";
        }
        double div = Math.floor(num.doubleValue() / base);
        String val = baseN(new BigDecimal(div), base);
        return StringUtils.stripStart(val, stripChars) + ALPHABET[num.remainder(new BigDecimal(base)).intValue()];
    }


    private Double getMaxIndex() {
        return (double) System.currentTimeMillis();
    }

    /**
     * This Method will check if dialcode has numeric value at odd indexes.
     *
     * @param code
     * @return Boolean
     */
    private Boolean isValidCode(String code) {
        Matcher matcher = pattern.matcher(code);
        return matcher.matches();
    }


}

File: certificate-processor/src/main/scala/org/sunbird/incredible/processor/qrcode/AccessCodeGenerator.scala
package org.sunbird.incredible.processor.qrcode


import scala.math.BigDecimal
import org.apache.commons.lang3.StringUtils

import scala.util.matching.Regex

class AccessCodeGenerator(length: Double = 6.0) {
  private val stripChars: String = "0"
  private val largePrimeNumber = BigDecimal(1679979167)
  private val regex: Regex = "[A-Z][0-9][A-Z][0-9][A-Z][0-9]".r
  private val pattern: java.util.regex.Pattern = regex.pattern

  private val ALPHABET = Array[String]("1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F", "G", "H", "J", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z")


  def generate(): String = {
    val count: Double = 1
    val totalChars: Int = ALPHABET.length
    var exponent: BigDecimal = BigDecimal.valueOf(totalChars)
    exponent = exponent.pow(length.intValue)
    var codesCount: Double = 0
    var lastIndex: Double = 0
    var code: String = null
    while (codesCount < count) {
      lastIndex = getMaxIndex
      val number: BigDecimal = BigDecimal(lastIndex)
      val num: BigDecimal = number.*(largePrimeNumber).remainder(exponent)
      code = baseN(num, totalChars)
      if (code.length == length && isValidCode(code))
        codesCount += 1
    }
    code
  }

  private def baseN(num: BigDecimal, base: Int): String = {
    if (num.doubleValue == 0) return "0"
    val div = Math.floor(num.doubleValue / base)
    val value = baseN(BigDecimal(div), base)
    StringUtils.stripStart(value, stripChars) + ALPHABET(num.remainder(BigDecimal(base)).intValue)
  }


  private def getMaxIndex = System.currentTimeMillis.toDouble

  /**
    * This Method will check if dialcode has numeric value at odd indexes.
    *
    * @param code
    * @return Boolean
    */
  private def isValidCode(code: String): Boolean = {
    val matcher = pattern.matcher(code)
    matcher.matches()
  }


}
